### PR TITLE
Throw error if data channel's buffer is filled, rather than closing.

### LINKED ISSUE #1148: How do applications know a DataChannel's buffer capacity, so they can avoid filling it?
A well-behaving application that sends a large amount of data will use the "bufferedAmount" event and "bufferedAmountLowThreshold", to ensure it doesn't fill the DataChannel's buffer and cause the channel to be closed.

But how does the application know what to set "bufferedAmountLowThreshold" to? If the spec doesn't provide any guarantees about an implementation's buffer size, the application developer is forced to research how large different browsers' buffers are, pick a value low enough, and hope the browser buffers don't get smaller.

So I'd suggest either:

* Providing a `bufferSize` attribute (pretty straightforward)
* Guarantee something about the buffer size. For example, "it MUST be N MB or larger".
* Get rid of the "data channel closing on buffer filling" behavior, and just throw/return an error instead, like a non-blocking socket. This is a more radical change, but is anyone really attached to the current behavior? It seems to cause a fair amount of confusion. See discussion here: https://groups.google.com/forum/#!topic/discuss-webrtc/apmq9VLY9lc

### PR DESCRIPTION
**Pull request by taylor-b**
Fixes #1148.

Option 2 for fixing the issue. The application doesn't need to know the
buffer's size ahead of time, because filling the buffer is no longer an
irrecoverable error. If it doesn't care about latency, an application
could send until an exception is thrown, then wait for the
"bufferedamountlow" event until it can send again, similar to a
non-blocking socket.

Note that this is deliberately different than how WebSockets work,
because the circumstances are slightly different: with WebSockets, it
was expected that applications had code to handle them being closed
unexpectedly. But this isn't true of WebRTC DataChannels, because before
an SCTP-level timeout can occur, the application would have done an ICE
restart and repaired the problem.

### CODE REVIEW

File: webrtc.html [hunk 1]
```diff
@@ -8941,11 +8941,16 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
...
+        </li>
+        <li>
           <p>Attempt to send <var>data</var> on <var>channel</var>'s
-          <a>underlying data transport</a>; if the data cannot be sent, e.g.
-          because it would need to be buffered but the buffer is full, the user
-          agent MUST abruptly <a>close</a> <var>channel</var>'s <a>underlying
-          data transport</a> <a>with an error</a>.</p>
+          <a>underlying data transport</a>. If the data cannot be sent, buffer
+          it, and if this is not possible because the buffer is full,
+          <a>throw</a> an <code>OperationError</code>.</p>
```
Comment (taylor-b) [inline webrtc.html]: I don't know if these are the right errors; suggestions welcome.

**Comment by stefhak** [comment]
Interim topic to decide between options? And are they exclusive?

**Comment by taylor-b** [comment]
They're not necessarily mutually exclusive, no.

**Comment by aboba** [comment]
Related to Issue https://github.com/w3c/webrtc-pc/issues/1287

**Comment by soareschen** [comment]
I think there are few clarifications needed:

- When there are already buffers queued, it should attempt to queue the buffer without attempting to send to underlying data transport.
- If there is error sending to underlying transport, should there be error thrown? Or should the interaction with underlying transport be asynchronous and queued as a task?

**Comment by taylor-b** [comment]
> - When there are already buffers queued, it should attempt to queue the buffer without attempting to send to underlying data transport.

Good catch... Reworded to make this clear.

> - If there is error sending to underlying transport, should there be error thrown? Or should the interaction with underlying transport be asynchronous and queued as a task?

It should always be asynchronous. That way the application only has one thing to listen to for SCTP protocol level errors: `onerror`.

> - If readyState is `closing` or `closed`, why is that error ignored while others are thrown?

I don't see how `send` could work if `readyState` is `"closing"` or `"closed"`. I changed the `readyState != "opening"` check to `readyState == "open"`.

**Comment by soareschen** [comment]
> I don't see how send could work if readyState is "closing" or "closed". I changed the readyState != "opening" check to readyState == "open".

:clap: 

LGTM

File: webrtc.html [hunk 2]
```diff
@@ -8941,11 +8941,20 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
...
-          data transport</a> <a>with an error</a>.</p>
+          <p>If the size of <var>data</var> exceeds the value of <code><a
+          data-link-for="RTCSctpTransport">maxMessageSize</a></code> on
+          <var>channel</var>'s associated <code>RTCSctpTransport</code>,
+          <a>throw</a> an <code>InvalidAccessError</code>.</p>
+        </li>
+        <li>
+          <p>Queue <var>data</var> for transmission on <var>channel</var>'s
+          <a>underlying data transport</a>. If queuing <var>data</var> is not
+          possible because no more buffer space is available, <a>throw</a> an
```
Comment (stefhak) [inline webrtc.html]: s/no/not enough/ ?

**Comment by stefhak** [comment]
Made a comment, but I think we can merge now (regardless to whether or not we change based on that comment).

File: webrtc.html [hunk 3]
```diff
@@ -8941,11 +8941,20 @@ <h2 id="sec.cert-mgmt">Certificate Management</h2>
...
         <li>
-          <p>Attempt to send <var>data</var> on <var>channel</var>'s
-          <a>underlying data transport</a>; if the data cannot be sent, e.g.
-          because it would need to be buffered but the buffer is full, the user
-          agent MUST abruptly <a>close</a> <var>channel</var>'s <a>underlying
-          data transport</a> <a>with an error</a>.</p>
+          <p>If the size of <var>data</var> exceeds the value of <code><a
+          data-link-for="RTCSctpTransport">maxMessageSize</a></code> on
+          <var>channel</var>'s associated <code>RTCSctpTransport</code>,
+          <a>throw</a> an <code>InvalidAccessError</code>.</p>
```
Comment (adam-be) [inline webrtc.html]: Perhaps we should go with an operation error here as well since InvalidAccessError is deprecated.

Comment (taylor-b) [inline webrtc.html]: Hmm. What's the new "invalid parameter" error, then?

File: webrtc.html (hunk 2 above)
Comment (taylor-b) [inline webrtc.html]: Done

**Comment by lgrahl** [comment]
We have added this in Firefox 57. I'm currently writing a shim in adapter-js. Any other implementations that support this I should to be aware of (preferrably with version number)?

**Pull request merged**